src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -413,11 +444,26 @@ void ZeroMQBackend::Run() {
     };
 
     auto HandleInprocMessages = [this](std::vector<MultipartMessage>& msgs) {
-        // Forward messages from the inprocess bridge to XSUB for subscription
-        // subscription handling (1 part) or XPUB for publishing (4 parts).
+        // Forward messages from the inprocess bridge.
+        //
+        // Either it's two parts (tag, payload) for controlling subscriptions
+        // or initiating a shutdown, or it is 4 parts in which case messages
+        // are directly forwarded to the XPUB socket.
         for ( auto& msg : msgs ) {
-            if ( msg.size() == 1 ) {
-                xsub.send(msg[0], zmq::send_flags::none);
+            if ( msg.size() == 2 ) {
+                InprocTag tag = msg[0].data<InprocTag>()[0];
+                if ( tag == InprocTag::XSUB ) {
+                    xsub.send(msg[1], zmq::send_flags::none);
+                }
+                else if ( tag == InprocTag::Shutdown ) {
+                    if ( self_thread_stop )
+                        ZEROMQ_THREAD_PRINTF("inproc: error: duplicate shutdown message");
+
+                    self_thread_stop = true;
+                }
+                else {
+                    ZEROMQ_THREAD_PRINTF("inproc: error: unknown tag %x", static_cast<uint8_t>(tag));
+                }
```

Comment:
Let's use a `switch` instead. Since the enum is declared in this file we know all possible values, and no `default` branch should be needed.

```suggestion
                switch ( msg[0].data<InprocTag>()[0] ) {
                    case InprocTag::XSUB: xsub.send(msg[1], zmq::send_flags::none); break;
                    case InprocTag::Shutdown:
                        if ( self_thread_stop )
                            ZEROMQ_THREAD_PRINTF("inproc: error: duplicate shutdown");
                        self_thread_stop = true;
                        break;
                }
```